domain/ide/workspace/folder_test.go

```diff
@@ -463,25 +465,71 @@ func Test_processResults_ShouldSendAnalyticsToAPI(t *testing.T) {
 		Issues:  []snyk.Issue{mockCodeIssue},
 	}
 
-	engineMock.EXPECT().GetConfiguration().AnyTimes().Return(gafConfig)
-	engineMock.EXPECT().InvokeWithInputAndConfig(localworkflows.WORKFLOWID_REPORT_ANALYTICS, gomock.Any(), gomock.Any()).
-		// this captures the call parameters of the mocked call
-		Do(func(id workflow.Identifier, workflowInputData []workflow.Data, config configuration.Configuration) {
-			require.Equal(t, 1, len(workflowInputData))
-			payloadBytes, ok := workflowInputData[0].GetPayload().([]byte)
-			require.True(t, ok)
-
-			var scanDoneEvent json_schemas.ScanDoneEvent
-			err := json.Unmarshal(payloadBytes, &scanDoneEvent)
+	ic := analytics.NewInstrumentationCollector()
+
+	ua := util.GetUserAgent(gafConfig, config.Version)
+	ic.SetUserAgent(ua)
+	categories := setupCategories(&data, c)
+	ic.SetCategory(categories)
+	ic.SetStage("dev")
+	ic.SetStatus("Success") //or get result status from scan
+	ic.SetInteractionType("Scan done")
+	summary := createTestSummary(&data, c)
+	ic.SetTestSummary(summary)
+	ic.SetType("Analytics")
+
+	entered := false
+	_, err := engineMock.Register(localworkflows.WORKFLOWID_REPORT_ANALYTICS, workflow.ConfigurationOptionsFromFlagset(pflag.NewFlagSet("", pflag.ContinueOnError)),
+		func(invocation workflow.InvocationContext, workflowInputData []workflow.Data) ([]workflow.Data, error) {
+			actualV2InstrumentationObject, err := analytics.GetV2InstrumentationObject(ic)
+
+			entered = true
 			require.NoError(t, err)
-			require.Equal(t, "Snyk Open Source", scanDoneEvent.Data.Attributes.ScanType)
-			require.Equal(t, 1, scanDoneEvent.Data.Attributes.UniqueIssueCount.Medium)
+
+			require.Equal(t, "snyk-ls", actualV2InstrumentationObject.Data.Attributes.Runtime.Application.Name)
+			require.Equal(t, "dev", string(*actualV2InstrumentationObject.Data.Attributes.Interaction.Stage))
+			require.Equal(t, "Success", actualV2InstrumentationObject.Data.Attributes.Interaction.Status)
+			require.Equal(t, "Scan done", actualV2InstrumentationObject.Data.Attributes.Interaction.Type)
+			require.Equal(t, []string{product.ToProductCodename(data.Product), "test"}, *actualV2InstrumentationObject.Data.Attributes.Interaction.Categories)
+			require.Equal(t, "Analytics", actualV2InstrumentationObject.Data.Type)
+
+			return nil, nil
 		})
+	if err != nil {
+		return
+	}
+
+	err = engineMock.Init()
+	if err != nil {
+		return
+	}
 
 	// Act
 	f.processResults(data)
-	// wait for async analytics sending
-	time.Sleep(time.Second)
+	maxWaitTime := 10 * time.Second
+	startTime := time.Now()
+	waitTime := 100 * time.Millisecond
+
+	for {
+		if entered {
+			break
+		}
+
+		time.Sleep(waitTime)
```

Comment:
Suggestion to make this a lot easier than a mutex... use a channel instead. Should have proposed this earlier!

Instead of a bool, declare a channel `entered := make(chan struct{})`.

At the bottom of the callback (after all the `require` checks there), close it.

At the end of the test, instead of a backoff delay, `select` on it:

```
select {
  <-entered:
  <-time.After(maxWaitTIme):
    t.Fatalf("timeout reached, callback was never called")
}
```